### PR TITLE
Fix positionMapping in stale data

### PR DESCRIPTION
**Pull request by soulomoon**
Fix https://github.com/haskell/haskell-language-server/issues/3919#issue-2054752246

### CODE REVIEW

**Comment by michaelpj** [comment]
I have absolutely no idea how this stuff works. Maybe @wz1000 does?

**Review by soulomoon — COMMENTED** [review]
the problem lays in `updatePositionMapping`.

**Comment by soulomoon** [comment]
To share the old keys being in acc, we should use `mapAccumWithKey`

**Comment by soulomoon** [comment]
I add tests on this. 
Just run `cabal test --test-option="-p /enumMapMappingTest/" `

1. Not many understand how is the bug a problem, the test case can serve as explaination.
2. The behaviour of the bug is subtle, it can prevent re-introduction. 

I will explain how this is going on,
Suppose you have a text starting from (0,0) `let x = 1 in 1`(version 0, with x in posision (0, 4)).
For i in [1..5], you are inserting a space at (0, i) consecutively.
After the 6th insertion, the text is now `     let x = 1 in 1`(version 6, with x in position (0, 9)).
The bug would case the mapping of (0, 4)(version0) to  (0, 8)(version5) which should be (0, 9)(version5)

It could be easily reproduced by swtich `addOldDelta` back to `addDelta` in `updatePositionMappingHelper` at /Users/ares/src/haskell-language-server/ghcide/src/Development/IDE/Core/Shake.hs


```
enumMapMappingTest :: TestTree
enumMapMappingTest = testCase "enumMapMappingTest" $ do
  let mkChangeEvent :: Range -> Text -> TextDocumentContentChangeEvent
      mkChangeEvent r t = TextDocumentContentChangeEvent $ InL $ #range .== r .+ #rangeLength .== Nothing .+ #text .== t
      mkCE :: UInt -> UInt -> UInt -> UInt -> Text -> TextDocumentContentChangeEvent
      mkCE l1 c1 l2 c2 = mkChangeEvent (Range (Position l1 c1) (Position l2 c2))
      events :: [(Int32, [TextDocumentContentChangeEvent])]
      events = map (second return) [(0, mkCE 0 0 0 0 ""), (1, mkCE 0 1 0 1 " "), (2, mkCE 0 2 0 2 " "), (3, mkCE 0 3 0 3 " "), (4, mkCE 0 4 0 4 " "), (5, mkCE 0 5 0 5 " ")]
      finalMap = Prelude.foldl (\m (i, e) -> updatePositionMappingHelper i e m) mempty events
  let updatePose fromPos = do
        mapping <- snd <$> EM.lookup 0 finalMap
        toCurrentPosition mapping fromPos
  updatePose (Position 0 4) @?= Just (Position 0 9)
  updatePose (Position 0 5) @?= Just (Position 0 10)
```

**Comment by soulomoon** [comment]
hi @michaelpj , I have added a test case, it should be easy to spot the bug now.

File: ghcide/src/Development/IDE/Core/PositionMapping.hs [hunk 1]
```diff
@@ -123,6 +124,10 @@ mkDelta cs = foldl' applyChange idDelta cs
 addDelta :: PositionDelta -> PositionMapping -> PositionMapping
 addDelta delta (PositionMapping pm) = PositionMapping (composeDelta delta pm)
 
+-- | Add a old delta onto a Mapping k n to make a Mapping (k - 1) n
```
Comment (wz1000) [inline ghcide/src/Development/IDE/Core/PositionMapping.hs:127]: Could you expand on the documentation for this as well as `addDelta` clarifying which is appropriate to use in which situation with something like the example you gave in the PR?

File: ghcide/src/Development/IDE/Core/PositionMapping.hs [hunk 2]
```diff
@@ -123,6 +124,10 @@ mkDelta cs = foldl' applyChange idDelta cs
 addDelta :: PositionDelta -> PositionMapping -> PositionMapping
 addDelta delta (PositionMapping pm) = PositionMapping (composeDelta delta pm)
```
Comment (wz1000) [inline ghcide/src/Development/IDE/Core/PositionMapping.hs:125]: Is `addDelta` used anywhere now?

**Review by wz1000 — CHANGES_REQUESTED** [review]
Thanks, great catch. All this is quite subtle!

File: ghcide/src/Development/IDE/Core/PositionMapping.hs (hunk 2 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/PositionMapping.hs:125]: No, not that I know of now.

Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/PositionMapping.hs:125]: Maybe we should delete addDelta?

Comment (wz1000) [reply ghcide/src/Development/IDE/Core/PositionMapping.hs:125]: Yes please

**Review by wz1000 — APPROVED** [review]

File: ghcide/src/Development/IDE/Core/PositionMapping.hs (hunk 1 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/PositionMapping.hs]: done, the added haddock comment should be clarifying things out.

**Comment by michaelpj** [comment]
Thanks so much for doing the work to add a test!

**Pull request merged**